src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingMiddleware.cs

```diff
@@ -60,58 +65,340 @@ public class ResponseCachingMiddleware
+
+                    // Update vary rules if they are different
+                    if (context.CachedVaryRules == null ||
+                        !StringValues.Equals(context.CachedVaryRules.Params, normalizedVaryParamsValue) ||
+                        !StringValues.Equals(context.CachedVaryRules.Headers, normalizedVaryHeaderValue))
+                    {
+                        context.CachedVaryRules = new CachedVaryRules
+                        {
+                            VaryKeyPrefix = FastGuid.NewGuid().IdString,
+                            Headers = normalizedVaryHeaderValue,
+                            Params = normalizedVaryParamsValue
+                        };
+
+                        context.Cache.Set(context.StorageBaseKey, context.CachedVaryRules, context.CachedResponseValidFor);
+                    }
+
+                    context.StorageVaryKey = context.KeyProvider.CreateStorageVaryKey(context);
+                }
+
+                // Ensure date header is set
+                if (context.TypedResponseHeaders.Date == null)
+                {
+                    context.TypedResponseHeaders.Date = context.ResponseTime;
+                }
+
+                // Store the response on the state
+                context.CachedResponse = new CachedResponse
+                {
+                    BodyKeyPrefix = FastGuid.NewGuid().IdString,
+                    Created = context.TypedResponseHeaders.Date.Value,
+                    StatusCode = context.HttpContext.Response.StatusCode
+                };
+
+                foreach (var header in context.TypedResponseHeaders.Headers)
+                {
+                    if (!string.Equals(header.Key, HeaderNames.Age, StringComparison.OrdinalIgnoreCase))
+                    {
+                        context.CachedResponse.Headers.Add(header);
+                    }
+                }
+            }
+            else
+            {
+                context.ResponseCacheStream.DisableBuffering();
+            }
+        }
+
+        internal static void FinalizeCachingBody(ResponseCachingContext context)
+        {
+            if (context.ShouldCacheResponse &&
+                context.ResponseCacheStream.BufferingEnabled &&
+                (context.TypedResponseHeaders.ContentLength == null ||
+                 context.TypedResponseHeaders.ContentLength == context.ResponseCacheStream.BufferedStream.Length))
+            {
+                if (context.ResponseCacheStream.BufferedStream.Length >= context.Options.MinimumSplitBodySize)
+                {
+                    // Store response and response body separately
+                    context.Cache.Set(context.StorageVaryKey ?? context.StorageBaseKey, context.CachedResponse, context.CachedResponseValidFor);
+
+                    var cachedResponseBody = new CachedResponseBody()
+                    {
+                        Body = context.ResponseCacheStream.BufferedStream.ToArray()
```

Comment:
This might result in super large byte arrays going to the the large object heap. We use a "paged" buffer type in Mvc to avoid creating large strings, so a similar strategy might be nice here:

* Current code's here: https://github.com/aspnet/Mvc/blob/df845236bc9f4d4266fc2b124fcd3ddaf7507dcf/src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/PagedBufferedTextWriter.cs
* Some refactoring that's in progress: https://github.com/aspnet/Mvc/blob/df845236bc9f4d4266fc2b124fcd3ddaf7507dcf/src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/PagedCharBuffer.cs